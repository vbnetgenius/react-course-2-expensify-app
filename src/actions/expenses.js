import { v4 as uuid }  from 'uuid';

// ADD_EXPENSE
export const addExpense = ({amount = '' , note = '', description = '', createdAt = ''} = {}) => ({
    type: 'ADD_EXPENSE',
    payload: {
        id: uuid(),
        amount,
        note,
        description,
        createdAt
    },
});

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    payload: {
        id
    },
});

// EDIT_EXPENSE
export const editExpense = ({id, ...rest}) => ({
    type: 'EDIT_EXPENSE',
    payload: {
        id,
        ...rest,
    },
});