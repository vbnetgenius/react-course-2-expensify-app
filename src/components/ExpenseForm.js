import React from 'react';
import { connect } from 'react-redux';

export class ExpenseForm extends React.Component {

    state = {
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense ? this.props.expense.amount : '',
        description: this.props.expense ? this.props.expense.description : '',
        error: undefined,    
    };

    handleAddExpense = (e) => {
        e.preventDefault();  
        
        if (!this.state.note || !this.state.amount || !this.state.description) {
            this.setState(() => ({
                error: 'Please fill out form fields',
            }));
        } else {
            
            this.setState(() => ({
                error: '',
            }));
                    
            this.props.onSubmit({
                note: this.state.note,
                amount: this.state.amount,
                description: this.state.description,                
            });

        }

    };

    handleChange = (e) => {
        this.setState((prevState) => {
            return {
                [e.target.name]: e.target.value,
            };
        });
    };

    render() {
        return <div>

            { this.state.error && <p>{this.state.error}</p> }

            <form onSubmit={this.handleAddExpense}>
                <input name="description" type="text" placeholder="Description" autoFocus={true} onChange={this.handleChange} />
                <input name="amount" type="text" placeholder="Amount" onChange={this.handleChange} />            
                <textarea name="note" cols="30" rows="10" onChange={this.handleChange}></textarea>           
                <button>Add Expense</button>
            </form>

        </div>
    }

}

export default connect()(ExpenseForm);