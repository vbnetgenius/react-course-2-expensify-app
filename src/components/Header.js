import React from 'react';
import { NavLink } from 'react-router-dom';


const Header = (props) => (
    <div>
        <h1>Siavash</h1>
        <NavLink to="/dashboard" activeClassName="bold">Dashboard</NavLink>
        <NavLink to="/profile" activeClassName="bold">Profile</NavLink>
        <NavLink to="/users" activeClassName="bold">Users</NavLink>
        <NavLink to="/about" activeClassName="bold">About</NavLink>
    </div>
);

export default Header;