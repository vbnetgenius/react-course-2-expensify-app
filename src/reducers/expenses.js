const expensedReducerDefaultState = [];
const expensesReducer = (state = expensedReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                {...action.payload}
            ];
        case 'REMOVE_EXPENSE':
            return state.filter(e => e.id != action.payload.id);
        case 'EDIT_EXPENSE':
            return state.map(e => {
                if (e.id === action.payload.id) {
                    return {
                        ...e,
                        ...action.payload,
                    };
                }
                return e;
            });
        default:
            return state;
    }
};

export default expensesReducer;