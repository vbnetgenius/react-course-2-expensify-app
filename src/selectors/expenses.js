// Get visible expenses
export default (expenses, { text, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate ;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = expense.description && expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    });
};