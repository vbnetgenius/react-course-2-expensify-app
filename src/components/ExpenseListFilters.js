import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from './../actions/filters';

const ExpenseListFilters = (props) => (
    <div>
        <input type="text" value={props.text} onChange={(e) => { props.dispatch(setTextFilter({text: e.target.value})) }} />
    </div>
);

const mapStateToProps = (state) => {
    return {
        text: state.filters.text,
    };
};

export default connect(mapStateToProps)(ExpenseListFilters);