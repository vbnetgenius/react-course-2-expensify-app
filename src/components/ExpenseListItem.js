import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

export const ExpenseListItem = (props) => {
    const { id, description, amount, createdAt } = props;
    return ( 
        <div>
            <h1>{description}</h1>
            <p>{amount} : {createdAt}</p>
            <button onClick={() => { props.dispatch( removeExpense( {id} ) ); } }>Delete</button>
        </div>
     );
}
 
export default connect()(ExpenseListItem);