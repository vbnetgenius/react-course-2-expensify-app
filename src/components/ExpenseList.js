import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import { v4 as uuid } from 'uuid';

export const ExpenseList = (props) => {    
    return <div>
    {
        props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map(e => <ExpenseListItem key={uuid()} {...e} />)
        )
    }
</div>;
}
 
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
    };
};

export default connect(mapStateToProps)(ExpenseList);