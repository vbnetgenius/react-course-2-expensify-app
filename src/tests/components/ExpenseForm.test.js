import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseForm } from './../../components/ExpenseForm';
import toJSON from 'enzyme-to-json';
import { expenses } from './../fixtures/expenses';

// test("should ExpenseForm render correctly", () => {
//     const wrapper = shallow(<ExpenseForm />);
//     expect(toJSON(wrapper)).toMatchSnapshot();
// });


test('should render form with error message', () => {
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(toJSON(wrapper)).toMatchSnapshot();
});


test('should set description on input change', () => {
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: {
            name: 'amount',
            value: 155000,
        },
    });
    expect(wrapper.state('amount')).toBe(155000);
});


test('should call onSubmit with valid form submissions', () => {
    const onSubmitSpy = jest.fn();
    const expense = expenses[0];
    const wrapper = shallow(<ExpenseForm expense={expense} onSubmit={onSubmitSpy} />);
    wrapper.find('form').first().simulate('submit', {
        preventDefault: () => {},
    });
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        note: expense.note,
        amount: expense.amount,
        description: expense.description
    });
});
