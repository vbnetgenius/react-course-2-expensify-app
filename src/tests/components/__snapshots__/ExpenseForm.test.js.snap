// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render form with error message 1`] = `
<div>
  <p>
    Please fill out form fields
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
    />
    <input
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="text"
    />
    <textarea
      cols="30"
      name="note"
      onChange={[Function]}
      rows="10"
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;
